lib/flipper/adapters/connection_pool.rb

```diff
@@ -0,0 +1,42 @@
+# An adapter that uses ConnectionPool to manage connections.
+#
+# Usage:
+#
+#   # Reuse an existing pool to create new adapters
+#   pool = ConnectionPool.new(size: 5, timeout: 5) { Redis.new }
+#   Flipper::Adapters::ConnectionPool.new(pool) do |client|
+#     Flipper::Adapters::Redis.new(client)
+#   end
+#
+#   # Create a new pool that returns the adapter
+#   Flipper::Adapters::ConnectionPool.new(size: 5, timeout: 5) do
+#     Flipper::Adapters::Redis.new(Redis.new)
+#   end
+class Flipper::Adapters::ConnectionPool
+  include Flipper::Adapter
+
+  # Use an existing ConnectionPool to initialize and cache new adapter instances.
+  class Wrapper
+    def initialize(pool, &initializer)
+      @pool = pool
+      @initializer = initializer
+      @instances = {}
```

Comment:
Could this hang onto resources for too long in some cases, preventing them from being garbage collected? `ConnectionPool` has a `reload` capability that will release its hold on the existing resources, but this `@instances` hash map will retain them and then grow larger as the pool offers up new resources.

I was initially thinking that `ObjectSpace::WeakKeyMap` would solve the problem, but this is a funny situation in which the map keys and values both have references to the same resource.

Reply:
Yes, good catch. I just pushed [18b2a1b](https://github.com/flippercloud/flipper/pull/835/commits/18b2a1bcbd6352a8d336a10d5dfa206e4af7f928) which makes `ConnectionPool` observable. I might open an issue on the uptream repo and see if a change like this would be accepted.

Reply:
That seems like a good solution. The only other thing that comes to mind is an LRU cache, but I don't know if there's a solid Ruby implementation.